pageserver/src/tenant/remote_layer.rs

```diff
@@ -0,0 +1,211 @@
+            .to_string()
+        })
+    }
+
+    fn local_path(&self) -> Option<PathBuf> {
+        None
+    }
+
+    fn get_value_reconstruct_data(
+        &self,
+        _key: Key,
+        _lsn_range: Range<Lsn>,
+        _reconstruct_state: &mut ValueReconstructState,
+    ) -> Result<ValueReconstructResult> {
+        bail!("layer {} needs to be downloaded", self.filename().display());
+    }
+
+    fn iter<'a>(
+        &'a self,
+    ) -> Result<Box<dyn Iterator<Item = anyhow::Result<(Key, Lsn, Value)>> + 'a>> {
+        bail!("cannot iterate a remote layer");
+    }
+
+    fn key_iter<'a>(&'a self) -> Result<Box<dyn Iterator<Item = (Key, Lsn, u64)> + 'a>> {
+        bail!("cannot iterate a remote layer");
+    }
+
+    fn delete(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn is_incremental(&self) -> bool {
+        self.is_incremental
+    }
+
+    fn is_in_memory(&self) -> bool {
+        false
+    }
+
+    /// debugging function to print out the contents of the layer
+    fn dump(&self, _verbose: bool) -> Result<()> {
+        println!(
+            "----- remote layer for ten {} tli {} keys {}-{} lsn {}-{} ----",
+            self.tenantid,
+            self.timelineid,
+            self.key_range.start,
+            self.key_range.end,
+            self.lsn_range.start,
+            self.lsn_range.end
+        );
+
+        Ok(())
+    }
+
+    fn downcast_remote_layer<'a>(self: Arc<Self>) -> Option<std::sync::Arc<RemoteLayer>> {
+        Some(self)
+    }
+
+    fn is_remote_layer(&self) -> bool {
+        true
+    }
+}
+
+impl RemoteLayer {
+    pub fn new_img(
+        tenantid: TenantId,
+        timelineid: TimelineId,
+        fname: &ImageFileName,
+    ) -> RemoteLayer {
+        let path =
+            RelativePath::from_local_path(&PathBuf::from(""), &PathBuf::from(fname.to_string()))
+                .unwrap();
+        RemoteLayer {
+            tenantid,
+            timelineid,
+            key_range: fname.key_range.clone(),
+            lsn_range: fname.lsn..(fname.lsn + 1),
+            download_watch: Mutex::new(None),
+            is_delta: false,
+            is_incremental: false,
+            path,
+        }
+    }
+
+    pub fn new_delta(
+        tenantid: TenantId,
+        timelineid: TimelineId,
+        fname: &DeltaFileName,
+    ) -> RemoteLayer {
+        let path =
+            RelativePath::from_local_path(&PathBuf::from(""), &PathBuf::from(fname.to_string()))
+                .unwrap();
+        RemoteLayer {
+            tenantid,
+            timelineid,
+            key_range: fname.key_range.clone(),
+            lsn_range: fname.lsn_range.clone(),
+            download_watch: Mutex::new(None),
+            is_delta: true,
+            is_incremental: true,
+            path,
+        }
+    }
+
+    /// Create a Layer struct representing this layer, after it has been downloaded.
+    pub fn create_downloaded_layer(&self, conf: &'static PageServerConf) -> Arc<dyn Layer> {
```

Comment:
I wonder, if any `pub async fn ...(self` method could be created to ensure we cannot use the old object and get the download correctly?